### PR TITLE
[OCPQE-18591] - Observe initial boot 'events' of the hosts with common patterns for UPI/IPI/Agent, generate JUnits and self-heal

### PR DESCRIPTION
**Pull request by bmanzari**
The idea of this story is to gather more information from the events happening during the installation steps, in particular from the first boot-up to the latest rebase of the nodes.

The idea is to exploit the Prow observers pattern: https://docs.ci.openshift.org/docs/internals/observer-pods/

The observers should work for both agent and BM IPI/UPI jobs, by identifying common patterns/events that can inform about the initial stages of the installation.

### CODE REVIEW

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-sno-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-sno-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-ha-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-disc-proxy-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-disc-proxy-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-disc-proxy-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-disc-proxy-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-disc-proxy-f7

File: ci-operator/step-registry/baremetal/lab/agent/install/baremetal-lab-agent-install-commands.sh [hunk 1]
```diff
@@ -230,6 +230,9 @@ echo -e "\nPreparing files for next steps in SHARED_DIR..."
 cp "${INSTALL_DIR}/auth/kubeconfig" "${SHARED_DIR}/"
 cp "${INSTALL_DIR}/auth/kubeadmin-password" "${SHARED_DIR}/"
 
+# Copy coreos stream file so the observer pod can check if the correct live image was booted
+oinst coreos print-stream-json > "${SHARED_DIR}/coreos-stream.json"
```
Comment (aleskandro) [inline ci-operator/step-registry/baremetal/lab/agent/install/baremetal-lab-agent-install-commands.sh:234]: can you add these two lines in the other install scripts (bm upi and bm ipi)? no tests needed for now, but having them is enough.

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-nightly-baremetal-compact-agent-ipv4-static-disc-proxy-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-nightly-baremetal-compact-agent-ipv4-dynamic-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

File: ci-operator/step-registry/baremetal/lab/upi/install/baremetal-lab-upi-install-commands.sh [hunk 2]
```diff
@@ -279,6 +279,9 @@ cp "${INSTALL_DIR}/metadata.json" "${SHARED_DIR}/"
 cp "${INSTALL_DIR}/auth/kubeconfig" "${SHARED_DIR}/"
 cp "${INSTALL_DIR}/auth/kubeadmin-password" "${SHARED_DIR}/"
 
+# Copy coreos stream file so the observer pod can check if the correct live image was booted
+oinst coreos print-stream-json > "${SHARED_DIR}/coreos-stream.json"
```
Comment (aleskandro) [inline ci-operator/step-registry/baremetal/lab/upi/install/baremetal-lab-upi-install-commands.sh:283]: Can you add this to the ipi step too?

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml [hunk 3]
```diff
@@ -1884,6 +1884,9 @@ tests:
       AUX_HOST: openshift-qe-metal-ci.arm.eng.rdu2.redhat.com
       RESERVE_BOOTSTRAP: "false"
       architecture: amd64
+    observers:
```
Comment (aleskandro) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:1887]: Can we add this at workflow/chain level or only here?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+  # shellcheck disable=SC2154
+  for bmhost in $(yq e -o=j -I=0 '.[]' "${HOSTS_FILE}"); do
+    ipmiRecord $bmhost &
+  done
+}
+
+
+function waitFor(){
+  local fileToWait="${1}"
+  while [ ! -f "${fileToWait}" ]; do
```
Comment (aleskandro) [inline ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:385]: Do you think a timeout is worth it?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+function checkNodes(){
+  # shellcheck disable=SC2154
+  for bmhost in $(yq e -o=j -I=0 '.[]' "${HOSTS_FILE}"); do
+    isNodeAlive "${bmhost}" &
+  done
+}
+
+function isPxeJob(){
+  if [[ $JOB_NAME == *"pxe"* ]]; then
```
Comment (aleskandro) [inline ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:358]:
```suggestion
  if [[ $JOB_NAME == *"-pxe-"* ]]; then
```

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+  fi
+  echo "Job name is $JOB_NAME , pxe? $IS_PXE_JOB"
+}
+
+function ipmiRecord(){
+      local bmhost="${1}"
+      . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
+      echo "SoL recording on ${bmc_address}"
+      ssh "${SSHOPTS[@]}" -t "root@${AUX_HOST}" << EOF > "${ARTIFACT_DIR}/${ip}_${name}_ipmi.txt"
+      ipmitool -I lanplus -H "$bmc_address" -U "$bmc_user" -P "$bmc_pass" -z 8196 sol activate usesolkeepalive &
```
Comment (aleskandro) [inline ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:369]: Does this work with arm64 HPE nodes?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+  waitFor $COREOS_STREAM_FILE
+  isPxeJob
+  recordIPMILog &
+  checkNodes
+}
+
+initObserverPod
+
+# Keep the observer pod alive for 1 hour
+sleep 3600
```
Comment (aleskandro) [inline ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:404]: why?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+  echo "Starting isNodeAlive for ${host}"
+  for i in $(seq 1 $MAX_RETRY); do
+    printf "%s: Checking SSH connectivity for %s %s/${MAX_RETRY}\n" "$(date --utc --iso=s)" "${ip}" "${i}"
+    status="$(isNodeReachable "$host" || true;)"
+    if [[ $status == $NODE_IS_REACHABLE ]]; then
+      echo "Node ${host} alive, waiting for services to come up..."
+      # journalRecord may not work even if SSH check passed
+      sleep 60
+      journalRecord $bmhost &
+      checkBootedImage "boot" $bmhost &
```
Comment (aleskandro) [inline ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:242]: do you ever wait for all these subprocesses to join?

**Comment by aleskandro** [comment]
Hi @bmanzari , I added some comments. One thing I've in mind is whether we can better define the Finite State Machine of each node and handle all the functions in a more sorted way. Currently, you start from one function at the end of the init, and that function initiates a very hard-to-follow chain of functions. I would simplify this chain, if possible... 

Another thing is that it looks like we create a bunch of subprocesses, but we don't wait for them to conclude (?)

I just got an initial read, I want to focus a bit more on this code

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 4 above)
Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:385]: timeout for what?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 8 above)
Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:242]: no, what's the purpose in doing so ?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 7 above)
Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:404]: probably not necessary

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 6 above)
Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:369]: it 'works' but there is no output from the nodes (both arm64 and x86)
https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/47240/rehearse-47240-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7/1766033600462983168/artifacts/baremetal-compact-agent-ipv4-static-connected-f7/observers-ocp-baremetal-qe/artifacts/192.168.83.75_master-02_ipmi.txt

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml (hunk 3 above)
Comment (bmanzari) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:1877]: testing it

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 6 above)
Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:369]: yeah, i don't think we can get output from that for the HPEs.

have a quick look at `cat $(which hpsol)` in the bastion for the HPE nodes.

For the Dell nodes, it is the well-known issue for the agent case, isn't it?

Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:369]: for dell nodes, when using ABI it stops working past bootloader screen, but as of now I can't even get the POST logs

Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:369]: this is probably due to the fact we are reaching IPMI behind a NAT.

Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:369]: I'm running ipmitool from the bastion host, aren't they on the same net segment?

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml (hunk 3 above)
Comment (bmanzari) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:1877]: tested, it works only there

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 6 above)
Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh]: oh yes, in that case, you shouldn't have the issue: SOL doesn't work with ipmitool in HPE servers, but Dell servers should still get at least up to grub.

Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh]: have you tried redirecting stderr too? `2>&1`?

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 8 above)
Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:242]: after you add the `wait` everywhere needed, any 'deadlocks'-like issues should become clearer (for example, the one you mentioned for the missing catch of the last reboot).

Try to have something like:

Main Process => One sub-process per Host => An infinite loop with a switch/case that call different functions based on the value of a $STATE variable.

The $STATE variable will track the evolution of an FSM that represents the lifecycle of the node.

If you still need to fork within a host's sub-process, you can do it in that switch/case and add a wait specific to the PID of the processes you need to join.

Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:242]: Also add better in-line comments to let future devs able to debug this script and follow the flow of the observers.

For me, it works in general, just need cleanup and handling the sub-processes lifecycle correctly

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-baremetal-compact-agent-ipv4-static-connected-f7

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh (hunk 8 above)
Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh]: is the `wait` in the `cleanup` function enough?

Comment (aleskandro) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh]: `wait` should be called everytime you instantiate N processes and want to wait for them to join. Also look at the parameters you can pass to wait on the version of bash we use: you might need to catch the PIDs of the subprocesses and then wait/join only a specific subset of them.

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-baremetal-pxe-compact-agent-ipv4-static-connected-f14

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by danilo-gemoli** [comment]
It's going to be way better once we merge https://github.com/openshift/ci-tools/pull/4149

**Comment by danilo-gemoli** [comment]
@bmanzari https://github.com/openshift/ci-tools/pull/4149 has merged, observers are now available at workflow level. Refactor you configuration and try the new feature out.

**Comment by bmanzari** [comment]
thanks, I'll do it today

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

**Comment by bmanzari** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-baremetal-sno-agent-ipv4-static-connected-fips-f7

File: ci-operator/step-registry/baremetal/lab/ipi/install/baremetal-lab-ipi-install-commands.sh [hunk 9]
```diff
@@ -188,6 +188,10 @@ cp "${INSTALL_DIR}/metadata.json" "${SHARED_DIR}/"
 cp "${INSTALL_DIR}/auth/kubeconfig" "${SHARED_DIR}/"
 cp "${INSTALL_DIR}/auth/kubeadmin-password" "${SHARED_DIR}/"
 
+# Creating file straight into $SHARED_DIR is not 100% reliable because of propagation issues (author guessing)
+oinst coreos print-stream-json > "${INSTALL_DIR}/coreos-stream.json"
+cp "${INSTALL_DIR}/coreos-stream.json" "${SHARED_DIR}/"
```
Comment (aleskandro) [inline ci-operator/step-registry/baremetal/lab/ipi/install/baremetal-lab-ipi-install-commands.sh:193]: Do we need this both in the INSTALL_DIR and SHARED_DIR or can we run this as in:

```suggestion
oinst coreos print-stream-json > "${SHARED_DIR}/coreos-stream.json"
```

**Review by aleskandro — COMMENTED** [review]
This pr is lgtm. Just one question about the junits. Did we validate them with the teams responsible for RP? 
Will they be imported correctly or do we need other chnages in the classifier?

**Comment by aleskandro** [comment]
@bmanzari do you mind fixing the script considering the errors higlighted by shellcheck  (try to avoid skipping rules as much as possible and use a separate commit to be sure you don't break this - amazing - work).

File: ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh [hunk 10]
```diff
@@ -0,0 +1,574 @@
+#!/bin/bash
+
+# shellcheck disable=SC2154
```
Comment (aleskandro) [inline ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:3]: Can we avoid suppressing them globally and even better supporting them by fixing the code? When they are really exception, we should comment and say why we are not respecting those rules

Comment (bmanzari) [reply ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh:3]: sure, I was testing global suppression for the exceptions like`. <(echo "$bmhost")` , I will fix the code

**Pull request closed without merging**